# New York Oil Spill Data: Dimensional Database Design and Analysis

## Part(1) & (2)
Part 1 – Creating your Database Objects
Using the agreed schema diagram from Part I, create the physical tables for all entities and their relationships. Make sure the table structures contain all primary, foreign, and unique keys (if applicable) as well as any default, and/or check constraints. The tables must also indicate the not null constraint for all applicable attributes. Provide comments on your tables and their attributes

Part 2 – Constructing your Database – Bulk Data Loading
Once you have created the tables and their constraints, construct (i.e., populate) the tables in bulk using the csv data files provided to you (e.g. attached California_Crime_2021_DB.zip).

### A. Creating and Loading the Database for NIBRS

__Creating the Database__
* Before creating the 'NIBRS' database, we use PostgreSQL's DROPDB command to remove the database if it already exists. As a habit you should drop database before you use it.

* Then, we create the database using the CREATEDB command.

* __Note:__ This is the format to create New Database: In this case the User is Student and the Database is called NIBRS, our database for this assignment.

In [145]:
!dropdb -U student NIBRS

dropdb: error: database removal failed: ERROR:  database "NIBRS" is being accessed by other users
DETAIL:  There is 1 other session using the database.


In [146]:
!createdb -U student NIBRS

createdb: error: database creation failed: ERROR:  database "NIBRS" already exists


__Loading the Database__ 

_Please note that the two lines of code below need to be run each time you  load your instance. It will load the Postgre SQL extension and allow you to login to your Database._

To load PostgreSQL, We need to:
* (1) Load the Extension and 
* (2) Call the Database using Postgre Query (i.e. Login to the Database)

In [147]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [148]:
%sql postgresql://student@/NIBRS

__Additional Note__

_Please set the name of your working directory to be within a folder called "Assignment2". Doing so, will allow our code to run smoothly on your end. Use the !pwd command to confirm that you are in the right directory._

In [149]:
!pwd

/home/ubuntu/notebooks/Assignment2


### B. Unzipping and Looking at the available CSV files

__Unzipping the Database CSV files___ 

Please Upload the California_Crime_2021_DB.zip file at this stage to your working directory to run the code.

Use the command line !pwd to confirm where to upload the Zip File.

In [150]:
!pwd

/home/ubuntu/notebooks/Assignment2


In [151]:
!unzip -o California_Crime_2021_DB.zip

Archive:  California_Crime_2021_DB.zip
  inflating: NIBRS_ARRESTEE.csv      
  inflating: NIBRS_DataDictionary_highlighted.pdf  
  inflating: NIBRS_ETHNICITY.csv     
  inflating: NIBRS_incident.csv      
  inflating: NIBRS_LOCATION_TYPE.csv  
  inflating: NIBRS_OFFENDER.csv      
  inflating: NIBRS_OFFENSE.csv       
  inflating: NIBRS_OFFENSE_TYPE.csv  
  inflating: NIBRS_RELATIONSHIP.csv  
  inflating: NIBRS_VICTIM.csv        
  inflating: NIBRS_VICTIM_OFFENDER_REL.csv  
  inflating: NIBRS_VICTIM_OFFENSE.csv  
  inflating: NIBRS_VICTIM_TYPE.csv   
  inflating: REF_RACE.csv            


We decided to remove the California_Crime_2021_DB.zip file to clean our working directory

In [152]:
!rm California_Crime_2021_DB.zip

In [153]:
!rm NIBRS_DataDictionary_highlighted.pdf 
#We don't need the Data Dictionary for any Queries so, we decided to delete in from our Working Directory

### C. Checking Data, Creating Tables and Loading Data
__Please Note:__ We referred to the Data Dictionary and Schema to identify which attributes/columns to select, create tables, and load data within those specific tables.

#### 1. Incident Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [154]:
!csvcut -n NIBRS_incident.csv

  1: data_year
  2: agency_id
  3: incident_id
  4: nibrs_month_id
  5: cargo_theft_flag
  6: submission_date
  7: incident_date
  8: report_date_flag
  9: incident_hour
 10: cleared_except_id
 11: cleared_except_date
 12: incident_status
 13: data_home
 14: orig_format
 15: did


##### b) Select the columns we need from csv file and append it to a new csv file named incident.csv

In [155]:
!pwd

/home/ubuntu/notebooks/Assignment2


In [156]:
!csvcut -c agency_id,incident_id,incident_date,incident_hour '/home/ubuntu/notebooks/Assignment2/NIBRS_incident.csv' > incident.csv

##### c) Preliminary look at the data and data types

In [157]:
!head -n 100 incident.csv | csvstat

  1. "agency_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         2
	Smallest value:        1516
	Largest value:         1823
	Sum:                   161443
	Mean:                  1630.737
	Median:                1516
	StDev:                 149.281
	Most common values:    1516 (62x)
	                       1823 (37x)

  2. "incident_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         99
	Smallest value:        133362971
	Largest value:         149760824
	Sum:                   13942342449
	Mean:                  140831741.909
	Median:                139098258
	StDev:                 5715933.64
	Most common values:    141774651 (1x)
	                       136500530 (1x)
	                       136580646 (1x)
	                       136530548 (1x)
	                       136516079 (1x)

  3. "incident_date"

	Type of data:          Date
	Contains null values:  False
	Unique values:         34
	Smallest value:  

##### d) Create the incident entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [158]:
%%sql
DROP TABLE IF EXISTS INCIDENT Cascade;

CREATE TABLE INCIDENT (
  agency_id     Numeric(10) NOT NULL,
  incident_id   Numeric(10) NOT NULL, 
  incident_date Date,
  incident_hour Numeric(2), 
  PRIMARY KEY (incident_id)
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### e) Loading the data from the new csv created above into the table

In [159]:
!pwd

/home/ubuntu/notebooks/Assignment2


In [160]:
%%sql
COPY INCIDENT FROM '/home/ubuntu/notebooks/Assignment2/incident.csv'
CSV
HEADER; 

 * postgresql://student@/NIBRS
118176 rows affected.


[]

##### f) Confirm whether the bulk loading is completed correctly

In [161]:
%%sql
Select * from INCIDENT
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('1516'), Decimal('141774651'), datetime.date(2021, 8, 26), Decimal('12')),
 (Decimal('1516'), Decimal('136500530'), datetime.date(2021, 1, 1), Decimal('0')),
 (Decimal('1516'), Decimal('136580646'), datetime.date(2021, 1, 1), Decimal('0')),
 (Decimal('1516'), Decimal('136530548'), datetime.date(2021, 1, 1), Decimal('1')),
 (Decimal('1516'), Decimal('136516079'), datetime.date(2021, 1, 1), Decimal('0')),
 (Decimal('1516'), Decimal('143243093'), datetime.date(2021, 1, 1), Decimal('1')),
 (Decimal('1516'), Decimal('137248920'), datetime.date(2021, 1, 1), Decimal('1')),
 (Decimal('1516'), Decimal('136694620'), datetime.date(2021, 1, 1), Decimal('2')),
 (Decimal('1516'), Decimal('136456227'), datetime.date(2021, 1, 1), Decimal('4')),
 (Decimal('1516'), Decimal('134186719'), datetime.date(2021, 1, 1), Decimal('2'))]

#### 2. Race Data
Reading CSV files, Creating Database Tables and Loading Data into Tables. 

We considered all the columns/attributes of the Race Entity the Entity wasn't specified within the Data Dictionary.

##### a) Looking at the csv file columns

In [162]:
!csvcut -n REF_RACE.csv

  1: race_id
  2: race_code
  3: race_desc
  4: sort_order
  5: start_year
  6: end_year
  7: notes


##### b) Preliminary look at the data and data types



In [163]:
!head -n 100 REF_RACE.csv | csvstat

  1. "race_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         12
	Smallest value:        10
	Largest value:         99
	Sum:                   603
	Mean:                  50.25
	Median:                42.5
	StDev:                 27.631
	Most common values:    98 (1x)
	                       10 (1x)
	                       20 (1x)
	                       30 (1x)
	                       40 (1x)

  2. "race_code"

	Type of data:          Text
	Contains null values:  False
	Unique values:         12
	Longest value:         2 characters
	Most common values:    U (1x)
	                       W (1x)
	                       B (1x)
	                       I (1x)
	                       A (1x)

  3. "race_desc"

	Type of data:          Text
	Contains null values:  False
	Unique values:         12
	Longest value:         49 characters
	Most common values:    Unknown (1x)
	                       White (1x)
	                       Black or African American (1

##### c) Create the race entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [164]:
%%sql
DROP TABLE IF EXISTS RACE Cascade;

CREATE TABLE RACE (
    race_id      Numeric(4)   NOT NULL,
    race_code    CHAR(2)      NOT NULL,  
    race_desc    VARCHAR(100) NOT NULL,
    sort_order   Numeric(2)   NOT NULL,
    start_year   Numeric(4),
    end_year     Numeric(4),
    notes        Varchar(100),
    PRIMARY KEY (race_id),
    UNIQUE (race_code) 
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### d) Load the data from csv file

In [165]:
!pwd

/home/ubuntu/notebooks/Assignment2


In [166]:
%%sql
COPY RACE FROM '/home/ubuntu/notebooks/Assignment2/REF_RACE.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
12 rows affected.


[]

##### e) Confirm whether the bulk loading is completed correctly

In [167]:
%%sql
Select * from RACE
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('98'), 'U ', 'Unknown', Decimal('98'), None, None, 'Race is explicitly unknown'),
 (Decimal('10'), 'W ', 'White', Decimal('10'), None, None, ''),
 (Decimal('20'), 'B ', 'Black or African American', Decimal('20'), None, None, ''),
 (Decimal('30'), 'I ', 'American Indian or Alaska Native', Decimal('30'), None, None, ''),
 (Decimal('40'), 'A ', 'Asian', Decimal('40'), Decimal('2013'), None, 'Includes Asian Indian'),
 (Decimal('41'), 'AP', 'Asian, Native Hawaiian, or Other Pacific Islander', Decimal('41'), Decimal('1980'), Decimal('2012'), 'Includes Asian Indian'),
 (Decimal('42'), 'C ', 'Chinese', Decimal('42'), Decimal('1960'), Decimal('1979'), ''),
 (Decimal('43'), 'J ', 'Japanese', Decimal('43'), Decimal('1960'), Decimal('1979'), ''),
 (Decimal('50'), 'P ', 'Native Hawaiian or Other Pacific Islander', Decimal('50'), Decimal('2013'), None, ''),
 (Decimal('60'), 'O ', 'Other', Decimal('60'), Decimal('1960'), Decimal('1979'), 'Includes Native Hawaiian or Other Pacific Islander and Asian Indian')]

#### 3. Ethnicity Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [168]:
!csvcut -n NIBRS_ETHNICITY.csv

  1: ethnicity_id
  2: ethnicity_code
  3: ethnicity_name


##### b) Preliminary look at the data and data types

In [169]:
!head -n 100 NIBRS_ETHNICITY.csv | csvstat

  1. "ethnicity_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         5
	Smallest value:        10
	Largest value:         50
	Sum:                   150
	Mean:                  30
	Median:                30
	StDev:                 15.811
	Most common values:    10 (1x)
	                       20 (1x)
	                       30 (1x)
	                       40 (1x)
	                       50 (1x)

  2. "ethnicity_code"

	Type of data:          Text
	Contains null values:  False
	Unique values:         5
	Longest value:         1 characters
	Most common values:    H (1x)
	                       N (1x)
	                       M (1x)
	                       U (1x)
	                       X (1x)

  3. "ethnicity_name"

	Type of data:          Text
	Contains null values:  False
	Unique values:         5
	Longest value:         22 characters
	Most common values:    Hispanic or Latino (1x)
	                       Not Hispanic or Latino (1x)
	                

##### c) Create the ethnicity entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [170]:
%%sql
DROP TABLE IF EXISTS ETHNICITY Cascade;

CREATE TABLE ETHNICITY (
  ethnicity_id     Numeric(4)   NOT NULL, 
  ethnicity_code   CHAR(1),      
  ethnicity_name   VARCHAR(100),
PRIMARY KEY (ethnicity_id),
UNIQUE (ethnicity_code)    
);    

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### d) Load the data from csv file

In [171]:
%%sql
COPY ETHNICITY FROM '/home/ubuntu/notebooks/Assignment2/NIBRS_ETHNICITY.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
5 rows affected.


[]

##### e) Confirm whether the bulk loading is completed correctly

In [172]:
%%sql
Select * from ETHNICITY
limit 10;

 * postgresql://student@/NIBRS
5 rows affected.


[(Decimal('10'), 'H', 'Hispanic or Latino'),
 (Decimal('20'), 'N', 'Not Hispanic or Latino'),
 (Decimal('30'), 'M', 'Multiple'),
 (Decimal('40'), 'U', 'Unknown'),
 (Decimal('50'), 'X', 'Not Specified')]

#### 4. Location Type Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [173]:
!csvcut -n NIBRS_LOCATION_TYPE.csv

  1: location_id
  2: location_code
  3: location_name


##### b) Preliminary look at the data and data types

In [174]:
!head -n 100 NIBRS_LOCATION_TYPE.csv | csvstat

  1. "location_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         47
	Smallest value:        1
	Largest value:         99
	Sum:                   1232
	Mean:                  26.213
	Median:                24
	StDev:                 20.057
	Most common values:    1 (1x)
	                       2 (1x)
	                       3 (1x)
	                       4 (1x)
	                       5 (1x)

  2. "location_code"

	Type of data:          Number
	Contains null values:  False
	Unique values:         47
	Smallest value:        0
	Largest value:         58
	Sum:                   1327
	Mean:                  28.234
	Median:                23
	StDev:                 19.072
	Most common values:    37 (1x)
	                       1 (1x)
	                       38 (1x)
	                       39 (1x)
	                       40 (1x)

  3. "location_name"

	Type of data:          Text
	Contains null values:  False
	Unique values:         47
	Longest value: 

##### c) Create the LOCATION_TYPE entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [175]:
%%sql
DROP TABLE IF EXISTS LOCATION_TYPE Cascade;

CREATE TABLE LOCATION_TYPE (
  location_id      Numeric(10) NOT NULL,   
  location_code    CHAR(2),   
  location_name    VARCHAR(100),
PRIMARY KEY (location_id),
UNIQUE (location_code)    
);   

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### d) Load the data from csv file

In [176]:
%%sql
COPY LOCATION_TYPE FROM '/home/ubuntu/notebooks/Assignment2/NIBRS_LOCATION_TYPE.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
47 rows affected.


[]

##### e) Confirm whether the bulk loading is completed correctly

In [177]:
%%sql
Select * from LOCATION_TYPE
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('1'), '37', 'Abandoned/Condemned Structure'),
 (Decimal('2'), '01', 'Air/Bus/Train Terminal'),
 (Decimal('3'), '38', 'Amusement Park'),
 (Decimal('4'), '39', 'Arena/Stadium/Fairgrounds/Coliseum'),
 (Decimal('5'), '40', 'ATM Separate from Bank'),
 (Decimal('6'), '41', 'Auto Dealership New/Used'),
 (Decimal('7'), '02', 'Bank/Savings and Loan'),
 (Decimal('8'), '03', 'Bar/Nightclub'),
 (Decimal('9'), '42', 'Camp/Campground'),
 (Decimal('10'), '04', 'Church/Synagogue/Temple/Mosque')]

#### 5. Offender Data 
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [178]:
!csvcut -n NIBRS_OFFENDER.csv

  1: data_year
  2: offender_id
  3: incident_id
  4: offender_seq_num
  5: age_id
  6: age_num
  7: sex_code
  8: race_id
  9: ethnicity_id
 10: age_range_low_num
 11: age_range_high_num


##### b) Select only the columns we need and append it to a csv file named offender.csv

In [179]:
!csvcut -c offender_id,incident_id,offender_seq_num,age_num,sex_code,race_id,ethnicity_id '/home/ubuntu/notebooks/Assignment2/NIBRS_OFFENDER.csv' > offender.csv

##### c) Preliminary look at the data and data types

In [180]:
!head -n 100 offender.csv | csvstat

  1. "offender_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         99
	Smallest value:        151542146
	Largest value:         170096007
	Sum:                   15672487728
	Mean:                  158307956.848
	Median:                155233089
	StDev:                 5951831.626
	Most common values:    161073930 (1x)
	                       155090892 (1x)
	                       155181712 (1x)
	                       155125626 (1x)
	                       155108544 (1x)

  2. "incident_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         72
	Smallest value:        133362971
	Largest value:         149760824
	Sum:                   13794288145
	Mean:                  139336243.889
	Median:                136624968
	StDev:                 5262333.455
	Most common values:    147282554 (9x)
	                       134859394 (4x)
	                       135775205 (4x)
	                       144963734 (3x)
	       

##### d) Dealing with the Non-Numeric Values in age_num column
__Note:__ From (c) we saw that there are some non-numeric values in the age_num column, so we decided to deal with them using the command line function. 

Below we used the command line to count the unique instances of NS (which we assumed meant that that specific Offender's Age was 'Not Supplied')

In [277]:
!csvcut -c age_num offender.csv | csvsort | uniq -c | sort -rn | head -n 10

/home/ubuntu/.local/lib/python3.8/site-packages/agate/table/from_csv.py:74: RuntimeWarning: Error sniffing CSV dialect: Could not determine delimiter
  47726 NS
   8051 00
   5529 30
   4136 25
   3238 35
   2635 40
   2634 27
   2468 32
   2248 28
   2200 29


__Note:__ We use awk function to find the non-numeric values NS etc (as seen above) and replace them with NULL values because:
* There are 47726 non-numeric value NS and if we drop their specific tuples/rows, our data integrity would be affected 

* The rows/tuples have primary keys that are foreign keys in other tables therefore, we didn't drop these rows/tuples to avoid future issues when loading data.

* NULL values are still useful and will allow us to do mathematical operations on the age_num column which is better than changing the column data type to CHAR. Futhermore, changing the datatype to CHAR would be going against the data dictionary, which we used as our guide in creating our table.

In [182]:
!awk -F, 'BEGIN {OFS = FS} { if ($4 ~ /^[a-zA-Z]+$/) $4 = NULL; print}' offender.csv > offender_updated.csv

##### e) Looking at the updated data and data types

In [183]:
!head -n 100 offender_updated.csv | csvstat

  1. "offender_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         99
	Smallest value:        151542146
	Largest value:         170096007
	Sum:                   15672487728
	Mean:                  158307956.848
	Median:                155233089
	StDev:                 5951831.626
	Most common values:    161073930 (1x)
	                       155090892 (1x)
	                       155181712 (1x)
	                       155125626 (1x)
	                       155108544 (1x)

  2. "incident_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         72
	Smallest value:        133362971
	Largest value:         149760824
	Sum:                   13794288145
	Mean:                  139336243.889
	Median:                136624968
	StDev:                 5262333.455
	Most common values:    147282554 (9x)
	                       134859394 (4x)
	                       135775205 (4x)
	                       144963734 (3x)
	       

##### f) Create the OFFENDER entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [184]:
%%sql
DROP TABLE IF EXISTS OFFENDER Cascade;

CREATE TABLE OFFENDER (
  offender_id          Numeric(10) NOT NULL,
  incident_id          Numeric(10) NOT NULL,
  offender_seq_num     Numeric(3), 
  age_num              NUmeric(3),  
  sex_code             CHAR(1),
  race_id              Numeric(4) NOT NULL,
  ethnicity_id         Numeric(4),
PRIMARY KEY (offender_id),
FOREIGN KEY (incident_id) REFERENCES INCIDENT(incident_id),
FOREIGN KEY (race_id) REFERENCES RACE(race_id),
FOREIGN KEY (ethnicity_id) REFERENCES ETHNICITY(ethnicity_id) 
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### g) Load the data from csv file

In [185]:
%%sql
COPY OFFENDER FROM '/home/ubuntu/notebooks/Assignment2/offender_updated.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
131342 rows affected.


[]

##### h) Confirm whether the bulk loading is completed correctly

In [186]:
%%sql
Select * from OFFENDER
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('161073930'), Decimal('141774651'), Decimal('1'), Decimal('33'), 'M', Decimal('10'), Decimal('20')),
 (Decimal('155090892'), Decimal('136500530'), Decimal('1'), Decimal('28'), 'M', Decimal('10'), Decimal('10')),
 (Decimal('155181712'), Decimal('136580646'), Decimal('1'), Decimal('36'), 'F', Decimal('10'), Decimal('20')),
 (Decimal('155125626'), Decimal('136530548'), Decimal('1'), Decimal('40'), 'F', Decimal('10'), Decimal('20')),
 (Decimal('155108544'), Decimal('136516079'), Decimal('1'), Decimal('30'), 'F', Decimal('30'), Decimal('20')),
 (Decimal('155108545'), Decimal('136516079'), Decimal('2'), Decimal('21'), 'M', Decimal('30'), Decimal('20')),
 (Decimal('162730171'), Decimal('143243093'), Decimal('1'), Decimal('27'), 'M', Decimal('10'), Decimal('10')),
 (Decimal('155943307'), Decimal('137248920'), Decimal('1'), Decimal('44'), 'F', Decimal('20'), Decimal('20')),
 (Decimal('155312639'), Decimal('136694620'), Decimal('1'), Decimal('29'), 'M', Decimal('10'), Decimal('10')),
 (Decimal('155039704'), Decimal('136456227'), Decimal('1'), Decimal('50'), 'M', Decimal('10'), Decimal('20'))]

#### 6. Offense Type Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [187]:
!csvcut -n NIBRS_OFFENSE_TYPE.csv

  1: offense_code
  2: offense_name
  3: crime_against
  4: ct_flag
  5: hc_flag
  6: hc_code
  7: offense_category_name
  8: offense_group


##### b) Select only the columns we need and append it to a csv file named offense_type.csv

In [188]:
!csvcut -c offense_code,offense_name,crime_against,offense_category_name '/home/ubuntu/notebooks/Assignment2/NIBRS_OFFENSE_TYPE.csv' > offense_type.csv

##### c) Preliminary look at the data and data types

In [189]:
!head -n 100 offense_type.csv | csvstat

  1. "offense_code"

	Type of data:          Text
	Contains null values:  False
	Unique values:         86
	Longest value:         3 characters
	Most common values:    09A (1x)
	                       09B (1x)
	                       09C (1x)
	                       11A (1x)
	                       11B (1x)

  2. "offense_name"

	Type of data:          Text
	Contains null values:  False
	Unique values:         86
	Longest value:         43 characters
	Most common values:    Murder and Nonnegligent Manslaughter (1x)
	                       Negligent Manslaughter (1x)
	                       Justifiable Homicide (1x)
	                       Rape (1x)
	                       Sodomy (1x)

  3. "crime_against"

	Type of data:          Text
	Contains null values:  False
	Unique values:         4
	Longest value:         11 characters
	Most common values:    Society (41x)
	                       Property (27x)
	                       Person (17x)
	                       Not a Crime (1x)

  4. 

##### d) Create the OFFENSE_TYPE entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)
Note: We use offense_code as the primary key because offense_type_id doesn't exist in the data. Also, offense code has non-numeric values so it has be created as a varchar attribute and change all references of victim_type_id in other tables to offense_code

In [190]:
%%sql
DROP TABLE IF EXISTS OFFENSE_TYPE Cascade;

CREATE TABLE OFFENSE_TYPE (
  offense_code                  VARCHAR(5)    NOT NULL,  
  offense_name                  VARCHAR(100)  NOT NULL,
  crime_against                 VARCHAR(100), 
  offense_category_name         VARCHAR(100),
PRIMARY KEY(offense_code)
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### e) Load the data from csv file

In [191]:
%%sql
COPY OFFENSE_TYPE FROM '/home/ubuntu/notebooks/Assignment2/offense_type.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
86 rows affected.


[]

##### f) Confirm whether the bulk loading is completed correctly

In [192]:
%%sql
Select * from OFFENSE_TYPE
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[('09A', 'Murder and Nonnegligent Manslaughter', 'Person', 'Homicide Offenses'),
 ('09B', 'Negligent Manslaughter', 'Person', 'Homicide Offenses'),
 ('09C', 'Justifiable Homicide', 'Person', 'Homicide Offenses'),
 ('11A', 'Rape', 'Person', 'Sex Offenses'),
 ('11B', 'Sodomy', 'Person', 'Sex Offenses'),
 ('11C', 'Sexual Assault With An Object', 'Person', 'Sex Offenses'),
 ('11D', 'Fondling', 'Person', 'Sex Offenses'),
 ('120', 'Robbery', 'Property', 'Robbery'),
 ('13A', 'Aggravated Assault', 'Person', 'Assault Offenses'),
 ('220', 'Burglary/Breaking & Entering', 'Property', 'Burglary/Breaking & Entering')]

#### 7. Offense Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [193]:
!csvcut -n NIBRS_OFFENSE.csv

  1: data_year
  2: offense_id
  3: incident_id
  4: offense_code
  5: attempt_complete_flag
  6: location_id
  7: num_premises_entered
  8: method_entry_code


##### b) Select only the columns we need and append it to a new csv file named offense.csv

In [194]:
!csvcut -c offense_id,incident_id,offense_code,location_id '/home/ubuntu/notebooks/Assignment2/NIBRS_OFFENSE.csv' > offense.csv

##### c) Preliminary look at the data and data types

In [195]:
!head -n 100 offense.csv | csvstat

  1. "offense_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         99
	Smallest value:        160652352
	Largest value:         179538866
	Sum:                   16326053622
	Mean:                  164909632.545
	Median:                164370837
	StDev:                 3979734.125
	Most common values:    170403568 (1x)
	                       164352006 (1x)
	                       164444800 (1x)
	                       164387076 (1x)
	                       164370837 (1x)

  2. "incident_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         78
	Smallest value:        133362971
	Largest value:         149760824
	Sum:                   13564512668
	Mean:                  137015279.475
	Median:                136516079
	StDev:                 3446662.877
	Most common values:    135691997 (4x)
	                       137248920 (3x)
	                       134011096 (3x)
	                       133522129 (3x)
	        

##### d) Create the OFFENSE entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)
 

In [196]:
%%sql
DROP TABLE IF EXISTS OFFENSE Cascade;

CREATE TABLE OFFENSE (
  offense_id          Numeric(10) NOT NULL,
  incident_id         Numeric(10) NOT NULL, 
  offense_code        VARCHAR(5)  NOT NULL, 
  location_id         Numeric(10) NOT NULL,  
PRIMARY KEY (offense_id),
FOREIGN KEY (incident_id) REFERENCES INCIDENT(incident_id),
FOREIGN KEY (offense_code) REFERENCES OFFENSE_TYPE(offense_code),
FOREIGN KEY (location_id) REFERENCES LOCATION_TYPE(location_id)
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### e) Load the data from csv file

In [197]:
%%sql
COPY OFFENSE FROM '/home/ubuntu/notebooks/Assignment2/offense.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
136848 rows affected.


[]

##### f) Confirm whether the bulk loading is completed correctly

In [198]:
%%sql
Select * from OFFENSE
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('170403568'), Decimal('141774651'), '11A', Decimal('35')),
 (Decimal('164352006'), Decimal('136500530'), '13A', Decimal('25')),
 (Decimal('164444800'), Decimal('136580646'), '13B', Decimal('35')),
 (Decimal('164387076'), Decimal('136530548'), '13B', Decimal('98')),
 (Decimal('164370837'), Decimal('136516079'), '13B', Decimal('14')),
 (Decimal('172082117'), Decimal('143243093'), '13B', Decimal('35')),
 (Decimal('165205693'), Decimal('137248920'), '280', Decimal('25')),
 (Decimal('165205694'), Decimal('137248920'), '35A', Decimal('25')),
 (Decimal('165205695'), Decimal('137248920'), '35B', Decimal('25')),
 (Decimal('164575617'), Decimal('136694620'), '13B', Decimal('35'))]

#### 8. Relationship Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [199]:
!csvcut -n NIBRS_RELATIONSHIP.csv

  1: relationship_id
  2: relationship_code
  3: relationship_name


##### b) Preliminary look at the data and data types

In [200]:
!head -n 100 NIBRS_RELATIONSHIP.csv | csvstat

  1. "relationship_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         27
	Smallest value:        1
	Largest value:         27
	Sum:                   378
	Mean:                  14
	Median:                14
	StDev:                 7.937
	Most common values:    1 (1x)
	                       2 (1x)
	                       3 (1x)
	                       4 (1x)
	                       5 (1x)

  2. "relationship_code"

	Type of data:          Text
	Contains null values:  False
	Unique values:         27
	Longest value:         2 characters
	Most common values:    AQ (1x)
	                       BE (1x)
	                       BG (1x)
	                       CF (1x)
	                       CH (1x)

  3. "relationship_name"

	Type of data:          Text
	Contains null values:  False
	Unique values:         27
	Longest value:         52 characters
	Most common values:    Victim Was Acquaintance (1x)
	                       Victim Was Babysittee (1x)
	  

##### c) Create the RELATIONSHIP entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [201]:
%%sql
DROP TABLE IF EXISTS RELATIONSHIP Cascade;

CREATE TABLE RELATIONSHIP (
  relationship_id          Numeric(4) NOT NULL,
  relationship_code        CHAR(2),
  relationship_name        VARCHAR(100),
PRIMARY KEY (relationship_id),
UNIQUE (relationship_code)
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### d) Load the data from csv file

In [202]:
%%sql
COPY RELATIONSHIP FROM '/home/ubuntu/notebooks/Assignment2/NIBRS_RELATIONSHIP.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
27 rows affected.


[]

##### e) Confirm whether the bulk loading is completed correctly

In [203]:
%%sql
Select * from RELATIONSHIP
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('1'), 'AQ', 'Victim Was Acquaintance'),
 (Decimal('2'), 'BE', 'Victim Was Babysittee'),
 (Decimal('3'), 'BG', 'Victim Was Boyfriend/Girlfriend'),
 (Decimal('4'), 'CF', 'Victim Was Child of Boyfriend or Girlfriend'),
 (Decimal('5'), 'CH', 'Victim Was Child'),
 (Decimal('6'), 'CS', 'Victim Was Common-Law Spouse'),
 (Decimal('7'), 'EE', 'Victim was Employee'),
 (Decimal('8'), 'ER', 'Victim was Employer'),
 (Decimal('9'), 'FR', 'Victim Was Friend'),
 (Decimal('10'), 'GC', 'Victim Was Grandchild')]

#### 9. Victim Type Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [204]:
!csvcut -n NIBRS_VICTIM_TYPE.csv

  1: victim_type_id
  2: victim_type_code
  3: victim_type_name


##### b) Preliminary look at the data and data types

In [205]:
!head -n 10000 NIBRS_VICTIM_TYPE.csv | csvstat

  1. "victim_type_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         9
	Smallest value:        1
	Largest value:         9
	Sum:                   45
	Mean:                  5
	Median:                5
	StDev:                 2.739
	Most common values:    1 (1x)
	                       2 (1x)
	                       3 (1x)
	                       4 (1x)
	                       5 (1x)

  2. "victim_type_code"

	Type of data:          Text
	Contains null values:  False
	Unique values:         9
	Longest value:         1 characters
	Most common values:    B (1x)
	                       F (1x)
	                       G (1x)
	                       I (1x)
	                       L (1x)

  3. "victim_type_name"

	Type of data:          Text
	Contains null values:  False
	Unique values:         9
	Longest value:         23 characters
	Most common values:    Business (1x)
	                       Financial Institution (1x)
	                       Governmen

##### c) Create the VICTIM_TYPE entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [206]:
%%sql
DROP TABLE IF EXISTS VICTIM_TYPE Cascade;

CREATE TABLE VICTIM_TYPE (
  victim_type_id     Numeric(4)   NOT NULL,
  victim_type_code   CHAR(1),
  victim_type_name   VARCHAR(100),
PRIMARY KEY (victim_type_id),
UNIQUE (victim_type_code)
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### d) Load the data from csv file

In [207]:
%%sql
COPY VICTIM_TYPE FROM '/home/ubuntu/notebooks/Assignment2/NIBRS_VICTIM_TYPE.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
9 rows affected.


[]

##### e) Confirm whether the bulk loading is completed correctly

In [208]:
%%sql
Select * from VICTIM_TYPE

 * postgresql://student@/NIBRS
9 rows affected.


[(Decimal('1'), 'B', 'Business'),
 (Decimal('2'), 'F', 'Financial Institution'),
 (Decimal('3'), 'G', 'Government'),
 (Decimal('4'), 'I', 'Individual'),
 (Decimal('5'), 'L', 'Law Enforcement Officer'),
 (Decimal('6'), 'O', 'Other'),
 (Decimal('7'), 'R', 'Religious Organization'),
 (Decimal('8'), 'S', 'Society/Public'),
 (Decimal('9'), 'U', 'Unknown')]

#### 10. Victim Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [209]:
!csvcut -n NIBRS_VICTIM.csv

  1: data_year
  2: victim_id
  3: incident_id
  4: victim_seq_num
  5: victim_type_id
  6: assignment_type_id
  7: activity_type_id
  8: outside_agency_id
  9: age_id
 10: age_num
 11: sex_code
 12: race_id
 13: ethnicity_id
 14: resident_status_code
 15: age_range_low_num
 16: age_code_range_high


##### b) Select only the columns we need

In [210]:
!csvcut -c victim_id,incident_id,victim_seq_num,victim_type_id,age_num,sex_code,race_id,ethnicity_id '/home/ubuntu/notebooks/Assignment2/NIBRS_VICTIM.csv' > victim.csv

##### c) Preliminary look at the csv data

In [211]:
!head -n 100 victim.csv | csvstat

  1. "victim_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         99
	Smallest value:        146510584
	Largest value:         165253072
	Sum:                   14891168948
	Mean:                  150415847.96
	Median:                149954107
	StDev:                 3911504.943
	Most common values:    156137715 (1x)
	                       150097155 (1x)
	                       150188607 (1x)
	                       150131489 (1x)
	                       150115100 (1x)

  2. "incident_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         80
	Smallest value:        133362971
	Largest value:         149760824
	Sum:                   13540477373
	Mean:                  136772498.717
	Median:                136374074
	StDev:                 3424259.353
	Most common values:    135691997 (6x)
	                       133522129 (4x)
	                       134183793 (3x)
	                       136432719 (3x)
	          

##### d) Dealing with non-numeric values in age_num
__Note:__ From (c) we saw that ther are some non-numeric values in the age_num column for the VICTIM Entity

In [212]:
!csvcut -c age_num victim.csv | csvsort | uniq -c | sort -r | head -n 10

/home/ubuntu/.local/lib/python3.8/site-packages/agate/table/from_csv.py:74: RuntimeWarning: Error sniffing CSV dialect: Could not determine delimiter
  38823 NS
   2643 00
   2548 26
   2547 30
   2517 29
   2478 28
   2476 31
   2468 25
   2438 27
   2375 32


Note: We use awk function to find the non-numeric values NS, NN, NP, BB etc (as seen above) and replace them with NULL values because:
* There are 38823 non-numeric value NS and if we drop them the data integrity would be affected 
* They have primary keys that are foreign keys in other tables and would create issues with loading data for those
* NUll values are still useful if we want to do mathematical operations on the age_num column rather than changing the column to text

In [213]:
!awk -F, 'BEGIN {OFS = FS} { if ($5 ~ /^[a-zA-Z]+$/) $5 = NULL; print}' victim.csv > victim_updated.csv

##### e) Looking at the updated data and data types

In [214]:
!head -n 100 victim_updated.csv | csvstat

  1. "victim_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         99
	Smallest value:        146510584
	Largest value:         165253072
	Sum:                   14891168948
	Mean:                  150415847.96
	Median:                149954107
	StDev:                 3911504.943
	Most common values:    156137715 (1x)
	                       150097155 (1x)
	                       150188607 (1x)
	                       150131489 (1x)
	                       150115100 (1x)

  2. "incident_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         80
	Smallest value:        133362971
	Largest value:         149760824
	Sum:                   13540477373
	Mean:                  136772498.717
	Median:                136374074
	StDev:                 3424259.353
	Most common values:    135691997 (6x)
	                       133522129 (4x)
	                       134183793 (3x)
	                       136432719 (3x)
	          

##### f) Create the VICTIM entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [215]:
%%sql
DROP TABLE IF EXISTS VICTIM Cascade;

CREATE TABLE VICTIM (
  victim_id       Numeric(10) NOT NULL, 
  incident_id     Numeric(10) NOT NULL,
  victim_seq_num  Numeric(3),
  victim_type_id  Numeric(4) NOT NULL,
  age_num         Numeric(3),
  sex_code        CHAR(1),
  race_id         Numeric(4) NOT NULL,
  ethnicity_id    Numeric(4),
PRIMARY KEY (victim_id),
FOREIGN KEY (incident_id) REFERENCES INCIDENT(incident_id),
FOREIGN KEY (race_id) REFERENCES RACE(race_id),
FOREIGN KEY (ethnicity_id) REFERENCES ETHNICITY(ethnicity_id),
FOREIGN KEY (victim_type_id) REFERENCES VICTIM_TYPE(victim_type_id)
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### g) Load the data from updated CSV file

In [216]:
%%sql
COPY VICTIM FROM '/home/ubuntu/notebooks/Assignment2/victim_updated.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
133987 rows affected.


[]

##### h) Confirm whether the bulk loading is completed correctly

In [217]:
%%sql
Select * from VICTIM
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('156137715'), Decimal('141774651'), Decimal('1'), Decimal('4'), Decimal('29'), 'F', Decimal('10'), Decimal('20')),
 (Decimal('150097155'), Decimal('136500530'), Decimal('1'), Decimal('4'), Decimal('17'), 'M', Decimal('10'), Decimal('10')),
 (Decimal('150188607'), Decimal('136580646'), Decimal('1'), Decimal('4'), Decimal('38'), 'M', Decimal('10'), Decimal('20')),
 (Decimal('150131489'), Decimal('136530548'), Decimal('1'), Decimal('4'), Decimal('42'), 'M', Decimal('20'), Decimal('20')),
 (Decimal('150115100'), Decimal('136516079'), Decimal('1'), Decimal('4'), Decimal('31'), 'F', Decimal('10'), Decimal('20')),
 (Decimal('150115101'), Decimal('136516079'), Decimal('2'), Decimal('4'), None, 'M', Decimal('10'), Decimal('20')),
 (Decimal('157809774'), Decimal('143243093'), Decimal('1'), Decimal('4'), Decimal('30'), 'F', Decimal('10'), Decimal('20')),
 (Decimal('150953725'), Decimal('137248920'), Decimal('1'), Decimal('4'), Decimal('46'), 'F', Decimal('10'), Decimal('20')),
 (Decimal('150953726'), Decimal('137248920'), Decimal('2'), Decimal('8'), None, 'X', Decimal('99'), Decimal('50')),
 (Decimal('150318415'), Decimal('136694620'), Decimal('1'), Decimal('4'), Decimal('20'), 'F', Decimal('10'), Decimal('10'))]

#### 11. Victim Offender Relationship Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [218]:
!csvcut -n NIBRS_VICTIM_OFFENDER_REL.csv

  1: data_year
  2: victim_id
  3: offender_id
  4: relationship_id
  5: nibrs_victim_offender_id


##### b) Select only the columns we need
Note: We didn't use the nirbs_victim_offender_id because the column was empty

In [219]:
!csvcut -c victim_id,offender_id,relationship_id '/home/ubuntu/notebooks/Assignment2/NIBRS_VICTIM_OFFENDER_REL.csv' > victim_offender_rel.csv

##### c) Preliminary look at the csv data and data types

In [220]:
!head -n 100 victim_offender_rel.csv | csvstat

  1. "victim_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         82
	Smallest value:        146510761
	Largest value:         162789351
	Sum:                   15167700178
	Mean:                  153209092.707
	Median:                150299870
	StDev:                 6153850.374
	Most common values:    162419294 (9x)
	                       149275901 (4x)
	                       162263507 (3x)
	                       150115100 (2x)
	                       150115101 (2x)

  2. "offender_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         89
	Smallest value:        151542325
	Largest value:         167657654
	Sum:                   15659050800
	Mean:                  158172230.303
	Median:                155293718
	StDev:                 6094132.387
	Most common values:    151726024 (4x)
	                       155012214 (3x)
	                       155108544 (2x)
	                       155108545 (2x)
	         

##### d) Create the VICTIM_OFFENDER_REL entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [221]:
%%sql
DROP TABLE IF EXISTS VICTIM_OFFENDER_REL Cascade;

CREATE TABLE VICTIM_OFFENDER_REL (
  victim_id           Numeric(10) NOT NULL, 
  offender_id         Numeric(10)  NOT NULL,
  relationship_id     Numeric(4)  NOT NULL,
PRIMARY KEY (victim_id,offender_id,relationship_id),
FOREIGN KEY (victim_id) REFERENCES VICTIM(victim_id),
FOREIGN KEY (offender_id) REFERENCES OFFENDER(offender_id),
FOREIGN KEY (relationship_id) REFERENCES RELATIONSHIP(relationship_id)    
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### e) Load the data from csv file

In [222]:
%%sql
COPY VICTIM_OFFENDER_REL FROM '/home/ubuntu/notebooks/Assignment2/victim_offender_rel.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
39187 rows affected.


[]

##### f) Confirm whether the bulk loading is completed correctly

In [223]:
%%sql
Select * from VICTIM_OFFENDER_REL
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('156137715'), Decimal('161073930'), Decimal('18')),
 (Decimal('150097155'), Decimal('155090892'), Decimal('1')),
 (Decimal('150188607'), Decimal('155181712'), Decimal('21')),
 (Decimal('150131489'), Decimal('155125626'), Decimal('4')),
 (Decimal('150115100'), Decimal('155108544'), Decimal('24')),
 (Decimal('150115100'), Decimal('155108545'), Decimal('24')),
 (Decimal('150115101'), Decimal('155108544'), Decimal('24')),
 (Decimal('150115101'), Decimal('155108545'), Decimal('24')),
 (Decimal('157809774'), Decimal('162730171'), Decimal('18')),
 (Decimal('150318415'), Decimal('155312639'), Decimal('4'))]

#### 12. Victim Offense Data
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [224]:
!csvcut -n NIBRS_VICTIM_OFFENSE.csv

  1: data_year
  2: victim_id
  3: offense_id


##### b) Select only the columns we need

In [225]:
!csvcut -c victim_id,offense_id '/home/ubuntu/notebooks/Assignment2/NIBRS_VICTIM_OFFENSE.csv' > victim_offense.csv

##### c) Preliminary look at the csv data and data types

In [226]:
!head -n 100 victim_offense.csv | csvstat

  1. "victim_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         86
	Smallest value:        146510584
	Largest value:         165253072
	Sum:                   14900549111
	Mean:                  150510597.081
	Median:                150097155
	StDev:                 4036769.726
	Most common values:    150953726 (2x)
	                       150234939 (2x)
	                       147262089 (2x)
	                       158089822 (2x)
	                       146737323 (2x)

  2. "offense_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         91
	Smallest value:        160652352
	Largest value:         179538866
	Sum:                   16307190441
	Mean:                  164719095.364
	Median:                164352006
	StDev:                 4083652.853
	Most common values:    161617992 (3x)
	                       160825859 (3x)
	                       164370837 (2x)
	                       161617991 (2x)
	          

##### d) Create the VICTIM_OFFENSE relationship table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [227]:
%%sql
DROP TABLE IF EXISTS VICTIM_OFFENSE Cascade;

CREATE TABLE VICTIM_OFFENSE (
  victim_id          Numeric(10) NOT NULL, 
  offense_id         Numeric(10)  NOT NULL,
PRIMARY KEY (victim_id,offense_id),
FOREIGN KEY (victim_id) REFERENCES VICTIM(victim_id),
FOREIGN KEY (offense_id) REFERENCES OFFENSE(offense_id)    
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### e) Load the data from csv file

In [228]:
%%sql
COPY VICTIM_OFFENSE FROM '/home/ubuntu/notebooks/Assignment2/victim_offense.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
148672 rows affected.


[]

##### f) Confirm whether the bulk loading is completed correctly

In [229]:
%%sql
Select * from VICTIM_OFFENSE
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('156137715'), Decimal('170403568')),
 (Decimal('150097155'), Decimal('164352006')),
 (Decimal('150188607'), Decimal('164444800')),
 (Decimal('150131489'), Decimal('164387076')),
 (Decimal('150115100'), Decimal('164370837')),
 (Decimal('150115101'), Decimal('164370837')),
 (Decimal('157809774'), Decimal('172082117')),
 (Decimal('150953725'), Decimal('165205693')),
 (Decimal('150953726'), Decimal('165205694')),
 (Decimal('150953726'), Decimal('165205695'))]

#### 13. Arrestee Data 
Reading CSV files, Creating Database Tables and Loading Data into Tables

##### a) Looking at the csv file columns

In [230]:
!csvcut -n NIBRS_ARRESTEE.csv

  1: data_year
  2: arrestee_id
  3: incident_id
  4: arrestee_seq_num
  5: arrest_date
  6: arrest_type_id
  7: multiple_indicator
  8: offense_code
  9: age_id
 10: age_num
 11: sex_code
 12: race_id
 13: ethnicity_id
 14: resident_code
 15: under_18_disposition_code
 16: clearance_ind
 17: age_range_low_num
 18: age_range_high_num


##### b) Select only the columns we need

In [231]:
!csvcut -c arrestee_id,incident_id,arrestee_seq_num,arrest_date,arrest_type_id,offense_code,age_num,sex_code,race_id,ethnicity_id '/home/ubuntu/notebooks/Assignment2/NIBRS_ARRESTEE.csv' > arrestee.csv

##### c) Preliminary look at the csv data and data types

In [232]:
!head -n 100 arrestee.csv | csvstat

  1. "arrestee_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         99
	Smallest value:        41190885
	Largest value:         44988200
	Sum:                   4194039304
	Mean:                  42364033.374
	Median:                42194293
	StDev:                 879020.153
	Most common values:    42182498 (1x)
	                       42171748 (1x)
	                       42375950 (1x)
	                       42215838 (1x)
	                       42148427 (1x)

  2. "incident_id"

	Type of data:          Number
	Contains null values:  False
	Unique values:         92
	Smallest value:        133363132
	Largest value:         146699111
	Sum:                   13590915117
	Mean:                  137281970.879
	Median:                136621605
	StDev:                 3053979.868
	Most common values:    138402584 (2x)
	                       138401524 (2x)
	                       133750424 (2x)
	                       135774219 (2x)
	                  

##### d) Create the ARRESTEE entity table based on the data types from csv and the needed rows from customer requirement (i.e. Data Dictionary)

In [233]:
%%sql
DROP TABLE IF EXISTS ARRESTEE Cascade;

CREATE TABLE ARRESTEE (
  arrestee_id          Numeric(10)   NOT NULL,
  incident_id          Numeric(10)   NOT NULL,
  arrestee_seq_num     Numeric(10),
  arrest_date          Date,
  arrest_type_id       Numeric(4),
  offense_code         VARCHAR(5)    NOT NULL,  
  age_num              Numeric(3), 
  sex_code             CHAR(1),
  race_id              Numeric(4)   NOT NULL,
  ethnicity_id         Numeric(4),   
PRIMARY KEY (arrestee_id),
FOREIGN KEY (incident_id) REFERENCES INCIDENT (incident_id),
FOREIGN KEY (offense_code) REFERENCES OFFENSE_TYPE(offense_code),
FOREIGN KEY (race_id) REFERENCES RACE (race_id),
FOREIGN KEY (ethnicity_id) REFERENCES ETHNICITY (ethnicity_id)    
);

 * postgresql://student@/NIBRS
Done.
Done.


[]

##### e) Load the data from CSV file

In [234]:
%%sql
COPY ARRESTEE FROM '/home/ubuntu/notebooks/Assignment2/arrestee.csv'
CSV
HEADER;

 * postgresql://student@/NIBRS
39567 rows affected.


[]

##### e) Confirm whether the bulk loading is completed correctly

In [235]:
%%sql
Select * from ARRESTEE
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('42182498'), Decimal('136580646'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '90Z', Decimal('36'), 'F', Decimal('10'), Decimal('50')),
 (Decimal('42171748'), Decimal('136530548'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '13B', Decimal('40'), 'F', Decimal('10'), Decimal('50')),
 (Decimal('42375950'), Decimal('137248920'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '280', Decimal('44'), 'F', Decimal('20'), Decimal('50')),
 (Decimal('42215838'), Decimal('136694620'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '13B', Decimal('29'), 'M', Decimal('10'), Decimal('50')),
 (Decimal('42148427'), Decimal('136456227'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '13B', Decimal('50'), 'M', Decimal('10'), Decimal('50')),
 (Decimal('42194211'), Decimal('136621397'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '35A', Decimal('56'), 'F', Decimal('10'), Decimal('50')),
 (Decimal('41446092'), Decimal('134174100'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('2'), '35B', Decimal('31'), 'M', Decimal('20'), Decimal('50')),
 (Decimal('43309689'), Decimal('140531006'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '13B', Decimal('23'), 'F', Decimal('10'), Decimal('50')),
 (Decimal('42495201'), Decimal('137690455'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '520', Decimal('53'), 'M', Decimal('10'), Decimal('50')),
 (Decimal('41379097'), Decimal('134011096'), Decimal('1'), datetime.date(2021, 1, 1), Decimal('1'), '13B', Decimal('25'), 'M', Decimal('10'), Decimal('50'))]

## Part (3)
For all tables, find the total number of rows loaded. Check your answer (using select queries) against the original text files (using Linux or csvkit commands) 

### Checking the data 
* We use 'SELECT COUNT' command to find the total number of rows loaded.
* Then, we use 'WC -l' command to count the total number of rows in the original text files. We also use 'TAIL' command to count only the rows, excluding the header.
* Finally, we compare the total row counts in the database tables to verify the data.

#### 1. Incident Data Verification 

In [236]:
%%sql 

SELECT COUNT(*) FROM INCIDENT

 * postgresql://student@/NIBRS
1 rows affected.


[(118176,)]

In [237]:
!tail -n +2 NIBRS_incident.csv | wc -l

118176


##### Result: We verified that the total number of rows loaded is equal to 118176, matching the total number of rows in the original text file

We decided to remove the NIBRS_incident.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'incident.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [238]:
!rm NIBRS_incident.csv

#### 2. Race Data Verification 

In [239]:
%%sql 

SELECT COUNT(*) FROM RACE

 * postgresql://student@/NIBRS
1 rows affected.


[(12,)]

In [240]:
!tail -n +2 REF_RACE.csv | wc -l

12


##### Result: We verified that the total number of rows loaded is equal to 12, matching the total number of rows in the original text file

#### 3. Ethnicity Data Verification 

In [241]:
%%sql 

SELECT COUNT(*) FROM ETHNICITY 

 * postgresql://student@/NIBRS
1 rows affected.


[(5,)]

In [242]:
!tail -n +2 NIBRS_ETHNICITY.csv | wc -l

5


##### Result: We verified that the total number of rows loaded is equal to 5, matching the total number of rows in the original text file

#### 4. Location Type Data Verification 

In [243]:
%%sql 

SELECT COUNT(*) FROM LOCATION_TYPE

 * postgresql://student@/NIBRS
1 rows affected.


[(47,)]

In [244]:
!tail -n +2 NIBRS_LOCATION_TYPE.csv | wc -l

47


##### Result: We verified that the total number of rows loaded is equal to 47, matching the total number of rows in the original text file

#### 5. Offender Data Verification 

In [245]:
%%sql 

SELECT COUNT(*) FROM OFFENDER

 * postgresql://student@/NIBRS
1 rows affected.


[(131342,)]

In [246]:
!tail -n +2 NIBRS_OFFENDER.csv | wc -l

131342


##### Result: We verified that the total number of rows loaded is equal to 131342, matching the total number of rows in the original text file

We decided to remove the NIBRS_OFFENDER.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'offense.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [247]:
!rm NIBRS_OFFENDER.csv

#### 6. Offense Type Data Verification 

In [248]:
%%sql 

SELECT COUNT(*) FROM OFFENSE_TYPE

 * postgresql://student@/NIBRS
1 rows affected.


[(86,)]

In [249]:
!tail -n +2 NIBRS_OFFENSE_TYPE.csv | wc -l

86


##### Result: We verified that the total number of rows loaded is equal to 86, matching the total number of rows in the original text file

We decided to remove the NIBRS_OFFENDER_TYPE.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'offense_type.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [250]:
!rm NIBRS_OFFENSE_TYPE.csv

#### 7. Offense Data Verification 

In [251]:
%%sql 

SELECT COUNT(*) FROM OFFENSE

 * postgresql://student@/NIBRS
1 rows affected.


[(136848,)]

In [252]:
!tail -n +2 NIBRS_OFFENSE.csv | wc -l

136848


##### Result: We verified that the total number of rows loaded is equal to 136848, matching the total number of rows in the original text file

We decided to remove the NIBRS_OFFENSE.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'offense.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [253]:
!rm NIBRS_OFFENSE.csv

#### 8. Relationship Data Verification 

In [254]:
%%sql 

SELECT COUNT(*) FROM RELATIONSHIP

 * postgresql://student@/NIBRS
1 rows affected.


[(27,)]

In [255]:
!tail -n +2 NIBRS_RELATIONSHIP.csv | wc -l

27


##### Result: We verified that the total number of rows loaded is equal to 27, matching the total number of rows in the original text file

#### 9. Victim Type Data Verification 

In [256]:
%%sql 

SELECT COUNT(*) FROM VICTIM_TYPE

 * postgresql://student@/NIBRS
1 rows affected.


[(9,)]

In [257]:
!tail -n +2 NIBRS_VICTIM_TYPE.csv | wc -l

9


##### Result: We verified that the total number of rows loaded is equal to 9, matching the total number of rows in the original text file

#### 10. Victim Data Verification 

In [258]:
%%sql 

SELECT COUNT(*) FROM VICTIM

 * postgresql://student@/NIBRS
1 rows affected.


[(133987,)]

In [259]:
!tail -n +2 NIBRS_VICTIM.csv | wc -l

133987


##### Result: We verified that the total number of rows loaded is equal to 133987, matching the total number of rows in the original text file

We decided to remove the NIBRS_VICTIM.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'victim.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [260]:
!rm NIBRS_VICTIM.csv

#### 11. Victim Offender Relationship Data Verification 

In [261]:
%%sql 
SELECT COUNT(*) FROM VICTIM_OFFENDER_REL

 * postgresql://student@/NIBRS
1 rows affected.


[(39187,)]

In [262]:
!tail -n +2 NIBRS_VICTIM_OFFENDER_REL.csv | wc -l

39187


##### Result: We verified that the total number of rows loaded is equal to 39187, matching the total number of rows in the original text file

We decided to remove the NIBRS_VICTIM_OFFENDER_REL.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'victim_offender_rel.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [263]:
!rm NIBRS_VICTIM_OFFENDER_REL.csv

#### 12. Victim_Offense Data Verification 

In [264]:
%%sql 

SELECT COUNT(*) FROM VICTIM_OFFENSE

 * postgresql://student@/NIBRS
1 rows affected.


[(148672,)]

In [265]:
!tail -n +2 NIBRS_VICTIM_OFFENSE.csv | wc -l

148672


##### Result: We verified that the total number of rows loaded is equal to 148672, matching the total number of rows in the original text file

We decided to remove the NIBRS_VICTIM_OFFENSE.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'victim_offense.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [266]:
!rm NIBRS_VICTIM_OFFENSE.csv

#### 13. Arrestee Data Verification 

In [267]:
%%sql 
SELECT COUNT(*) FROM ARRESTEE

 * postgresql://student@/NIBRS
1 rows affected.


[(39567,)]

In [268]:
!tail -n +2 NIBRS_ARRESTEE.csv | wc -l

39567


##### Result: We verified that the total number of rows loaded is equal to 39567, matching the total number of rows in the original text file

We decided to remove the NIBRS_ARRESTEE.csv file using the !rm command to clean our working directory, since we don't need it any longer. The selected columns, we need are in a new csv file named 'arrestee.csv'. This is the file we used to make our table and run PostgreSQL Queries with.

In [269]:
!rm NIBRS_ARRESTEE.csv

## Part (4)

### Part 4A
#### Write a query to determine the top 5 offenses and bottom 5 offenses in terms of the total number of offenses committed. In the query provide the Offense code, offense name, Offense category, and the number of offenses.

##### Top 5 Type of Offenses

In [270]:
%%sql

SELECT 
    O.OFFENSE_CODE as "Offense Code", 
    OT.OFFENSE_NAME as "Offense Name", 
    OT.OFFENSE_CATEGORY_NAME as "Offense Category", 
    count(*) number_of_offenses
FROM 
    OFFENSE O, 
    OFFENSE_TYPE OT
WHERE 
    O.OFFENSE_CODE = OT.OFFENSE_CODE -- Offense Code is a PK in OFFENSE and FK in OFFENSE_TYPE Entity
GROUP BY 
    O.OFFENSE_CODE,
    OT.OFFENSE_NAME, 
    OT.OFFENSE_CATEGORY_NAME
ORDER BY 
    number_of_offenses DESC
limit 5;

 * postgresql://student@/NIBRS
5 rows affected.


[('13B', 'Simple Assault', 'Assault Offenses', 17800),
 ('290', 'Destruction/Damage/Vandalism of Property', 'Destruction/Damage/Vandalism of Property', 14837),
 ('23F', 'Theft From Motor Vehicle', 'Larceny/Theft Offenses', 14521),
 ('35A', 'Drug/Narcotic Violations', 'Drug/Narcotic Offenses', 13868),
 ('23H', 'All Other Larceny', 'Larceny/Theft Offenses', 11535)]

##### Bottom 5 Type of Offenses

In [271]:
%%sql

SELECT 
    O.OFFENSE_CODE as "Offense Code", 
    OT.OFFENSE_NAME as "Offense Name", 
    OT.OFFENSE_CATEGORY_NAME as "Offense Category", 
    count(*) number_of_offenses
FROM 
    OFFENSE O, 
    OFFENSE_TYPE OT
WHERE 
    O.OFFENSE_CODE = OT.OFFENSE_CODE -- Offense Code is a PK in OFFENSE and FK in OFFENSE TYPE Entity
GROUP BY 
    O.OFFENSE_CODE,
    OT.OFFENSE_NAME, 
    OT.OFFENSE_CATEGORY_NAME
ORDER BY 
    number_of_offenses
limit 5;

 * postgresql://student@/NIBRS
5 rows affected.


[('39B', 'Operating/Promoting/Assisting Gambling', 'Gambling Offenses', 1),
 ('49A', 'Harboring Escapee/Concealing from Arrest', 'Other Offenses', 1),
 ('09B', 'Negligent Manslaughter', 'Homicide Offenses', 1),
 ('39A', 'Betting/Wagering', 'Gambling Offenses', 1),
 ('09C', 'Justifiable Homicide', 'Homicide Offenses', 2)]

### Part 4B 
#### Write a query to find the distribution of victims by sex and race and offense location. Only include victims of type “Individual”. In the query, provide the sex, race, location, and count of offenses

Note: We put a limit of 10 on the ouput of the query. To see more rows for the same query, change the limit in the last row

In [272]:
%%sql

SELECT 
    V.sex_code Sex,
    R.race_desc Race,
    L.location_name Offense_Location,
    count(*) count_of_offenses
FROM 
    VICTIM V,
    VICTIM_TYPE VT,
    VICTIM_OFFENSE VO,
    OFFENSE O,
    LOCATION_TYPE L,
    RACE R
WHERE
    V.victim_type_id = VT.victim_type_id and -- Victim Type ID is a PK in Victim Type and a FK in Victim Entity
    VT.victim_type_name = 'Individual' and -- Selecting the Victim Type: "Individual"
    V.victim_id = VO.victim_id and -- Victim ID is a PK in Victim and a FK in Victim Offense Entity
    VO.offense_id = O.offense_id and -- Offense ID is a PK in Offense and a FK in Victim Offense Entity
    O.location_id = L.location_id and -- Location ID is a PK in Location Type and a FK in OFFENSE Entity
    V.race_id = R.race_id -- Race ID is a PK in Race and a FK in Victim Entity
GROUP BY 
    Sex,
    Race,
    Offense_Location
ORDER BY
    count_of_offenses DESC
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[('F', 'White', 'Residence/Home', 15376),
 ('M', 'White', 'Residence/Home', 13481),
 ('M', 'White', 'Highway/Road/Alley/Street/Sidewalk', 8975),
 ('F', 'White', 'Highway/Road/Alley/Street/Sidewalk', 5202),
 ('M', 'White', 'Parking/Drop Lot/Garage', 5078),
 ('F', 'White', 'Parking/Drop Lot/Garage', 3728),
 ('M', 'White', 'Park/Playground', 2167),
 ('M', 'Unknown', 'Highway/Road/Alley/Street/Sidewalk', 2158),
 ('F', 'Black or African American', 'Residence/Home', 2035),
 ('M', 'White', 'Commercial/Office Building', 1539)]

### Part 4C 
### Write a query to provide a list of victims who suffered at least 4 offenses in an incident. For each victim, provide victim id, sex, race, ethnicity, and location.

* First we join the tables of Victim and Victim_Offense on common victim_id
* Then we join them with Offense table based on offense_id from victim_offense and offense tables
* Then we match victims incident_id to offense's victim_id to make sure they are for the same incident
* We join it with offense using incident_id because we need to access the loactions for each of the offenses
* We join the above with race using race_id
* Then join it with ethnicity using ethnicity_id
* Since we are aggregating the names of offense_locations into one column and counting we group by all the columns in the select statement
* We filter based on count >= 4 to get victims that suffered 4 or more offenses
* Finally we ordered the query output by count_of_offenses
* In the end we get the victim_id, sex, race, ethnicity, count of offenses suffered and their locations for victims that suffered 4 or more offenses

Note: We limited the output of the query to 10, but if you want to see all please remove the limit in the last row

In [273]:
%%sql
SELECT
    V.victim_id,
    V.sex_code Sex,
    R.race_desc Race,
    E.ethnicity_name Ethnicity,
    count(*) count_of_offenses,
    STRING_AGG(L.location_name,' | ') Locations_of_offenses
FROM
    VICTIM V,
    RACE R,
    ETHNICITY E,
    OFFENSE O,
    VICTIM_OFFENSE VO,
    LOCATION_TYPE L
WHERE
    V.victim_id = VO.victim_id -- Victim_ID is a PK in Victim and a FK in Victim Offense Entity
    and
    VO.offense_id = O.offense_id -- Offense_ID is a PK in Offense and a FK in Victim Offense Entity
    and
    V.incident_id = O.incident_id -- Incident is a FK in Victim and a FK in Offense Entity, We want the location ID so we use Incident ID to find it
    and
    O.location_id = L.location_id -- Location_ID is a PK in Location Type and a FK in Offense
    and
    V.ethnicity_id = E.ethnicity_id -- Ethnicity ID is a PK in Ethnicity and a FK in Victim
    and
    V.race_id = R.race_id -- Race ID is a PK in Race and a FK in Victim Entity
GROUP BY
    V.victim_id,
    Sex,
    Race,
    Ethnicity
HAVING
    count(*) >= 4
ORDER BY
   count_of_offenses DESC
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('157802745'), 'F', 'Black or African American', 'Not Hispanic or Latino', 5, 'Park/Playground | Parking/Drop Lot/Garage | Parking/Drop Lot/Garage | Parking/Drop Lot/Garage | Parking/Drop Lot/Garage'),
 (Decimal('148279702'), 'F', 'White', 'Not Hispanic or Latino', 4, 'Specialty Store | Parking/Drop Lot/Garage | Specialty Store | Parking/Drop Lot/Garage'),
 (Decimal('151353643'), 'F', 'White', 'Not Hispanic or Latino', 4, 'Residence/Home | Residence/Home | Residence/Home | Residence/Home'),
 (Decimal('149952429'), 'M', 'White', 'Not Hispanic or Latino', 4, 'Residence/Home | Residence/Home | Residence/Home | Residence/Home'),
 (Decimal('150577511'), 'X', 'Not Specified', 'Not Specified', 4, 'Hotel/Motel/Etc. | Hotel/Motel/Etc. | Hotel/Motel/Etc. | Hotel/Motel/Etc.'),
 (Decimal('150927896'), 'F', 'White', 'Hispanic or Latino', 4, 'Residence/Home | Residence/Home | Residence/Home | Residence/Home'),
 (Decimal('149952428'), 'F', 'White', 'Not Hispanic or Latino', 4, 'Residence/Home | Residence/Home | Residence/Home | Residence/Home'),
 (Decimal('151933503'), 'M', 'White', 'Not Hispanic or Latino', 4, 'Hotel/Motel/Etc. | Hotel/Motel/Etc. | Hotel/Motel/Etc. | Park/Playground'),
 (Decimal('151945322'), 'F', 'White', 'Not Hispanic or Latino', 4, 'Highway/Road/Alley/Street/Sidewalk | Highway/Road/Alley/Street/Sidewalk | Park/Playground | Highway/Road/Alley/Street/Sidewalk'),
 (Decimal('147189631'), 'M', 'Asian', 'Not Hispanic or Latino', 4, 'Residence/Home | Residence/Home | Residence/Home | Residence/Home')]

### Part 4D
#### Write a query to provide a list of incidents with more than 9 victims. Provide a distribution of the victims (counts) based on their sex, race, and ethnicity.

#### Step 1: We create a view to query the incidents that had more than 9 victims
* First we join the tables of Victim and incident on incident_id
* After that we group by incident_id to count the total number of times a victims associated with each incident
* We limit our results of victims to more than 9 by using restriction on the count
* We get a table/view that gives the victim_id for victims and the count of number of victims they had

In [274]:
%%sql
CREATE OR REPLACE VIEW INCIDENTS_WITH_MORE_THAN_9_VICTIMS
AS
SELECT 
    I.incident_id,
    count(*)
FROM 
    INCIDENT I,
    VICTIM V
WHERE 
    I.incident_id=V.incident_id -- Incident ID is a PK in Incident and a FK in Victim Entity
GROUP BY
    I.incident_id    
HAVING 
    count(*) > 9;

 * postgresql://student@/NIBRS
Done.


[]

In [275]:
%%sql

SELECT incident_id, count FROM INCIDENTS_WITH_MORE_THAN_9_VICTIMS
limit 10;

 * postgresql://student@/NIBRS
10 rows affected.


[(Decimal('133449299'), 12),
 (Decimal('133519796'), 14),
 (Decimal('133601291'), 19),
 (Decimal('133673209'), 12),
 (Decimal('133750424'), 15),
 (Decimal('133948657'), 10),
 (Decimal('134105803'), 30),
 (Decimal('134557887'), 13),
 (Decimal('134920194'), 31),
 (Decimal('134926087'), 13)]

#### Step 2: We use the view we created above to filter incidents and join them to ethnicity and race tables to get the final form of the query
* We use the view created for querying incidents with more than 9 victims
* We join the above with race using race_id
* Then join it with ethnicity using ethnicity_id
* We group by incident_id, sex_code, ethnicity_name to aggregate the distribution of these incidents

In [276]:
%%sql

SELECT 
    V.sex_code Sex,
    R.race_desc Race,
    E.ethnicity_name Ethnicity,
    count(*) no_of_victims
FROM 
    VICTIM V,
    RACE R,
    ETHNICITY E
WHERE 
    V.incident_id IN (SELECT incident_id FROM INCIDENTS_WITH_MORE_THAN_9_VICTIMS)
    and
    V.race_id=R.race_id -- Race ID is a PK in Race and a FK in Victim Entity
    and
    V.ethnicity_id=E.ethnicity_id -- Ethnicity ID is a PK in Ethnicity and a FK in Victim Entity
    
GROUP BY 
    Sex,
    Race,
    Ethnicity
ORDER BY
    no_of_victims DESC
;

 * postgresql://student@/NIBRS
19 rows affected.


[('U', 'Unknown', 'Not Specified', 492),
 ('M', 'Unknown', 'Not Specified', 376),
 ('F', 'Unknown', 'Not Specified', 302),
 ('M', 'White', 'Not Hispanic or Latino', 193),
 ('X', 'Not Specified', 'Not Specified', 166),
 ('M', 'White', 'Hispanic or Latino', 156),
 ('F', 'White', 'Not Hispanic or Latino', 138),
 ('F', 'White', 'Hispanic or Latino', 85),
 ('M', 'Asian', 'Not Hispanic or Latino', 29),
 ('M', 'Black or African American', 'Not Hispanic or Latino', 26),
 ('F', 'Black or African American', 'Not Hispanic or Latino', 24),
 ('F', 'Asian', 'Not Hispanic or Latino', 23),
 ('M', 'Unknown', 'Unknown', 12),
 ('F', 'Unknown', 'Unknown', 12),
 ('M', 'Native Hawaiian or Other Pacific Islander', 'Not Hispanic or Latino', 3),
 ('U', 'Asian', 'Not Hispanic or Latino', 2),
 ('F', 'Native Hawaiian or Other Pacific Islander', 'Not Hispanic or Latino', 1),
 ('U', 'White', 'Hispanic or Latino', 1),
 ('F', 'American Indian or Alaska Native', 'Not Hispanic or Latino', 1)]